src/bidi-engine.js

```diff
@@ -0,0 +1,473 @@
+		} else if ( range === 0xFC || range === 0xFD ) {
+			return "AL";
+		} else if ( _LTR_RANGES_REG_EXPR.test( range ) ) { //unlikely case
+			return "L";
+		}  else if ( range === 8 ) { // even less likely
+			return "R";
+		}
+		return "N"; //undefined type, mark as neutral
+	};
+
+	var _isContextualDirRtl = function( text ) {
+		for ( var i = 0, charType; i < text.length; i++ ) {
+			charType = _getCharType( text.charAt( i ) );
+			if ( charType === "L" ) {
+				return false;
+			} else if ( charType === "R" ) {
+				return true;
+			}
+		}
+		return false;
+	};
+
+	var _resolveCharType = function( chars, types, resolvedTypes, idx ) {
+		var cType = types[idx], wType, nType, len, i;
+		switch ( cType ) {
+			case "L":
+			case "R":
+				_lastArabic = false;
+				break;
+			case "N":
+			case "AN":
+				break;
+
+			case "EN":
+				if ( _lastArabic ) {
+					cType = "AN";
+				}
+				break;
+
+			case "AL":
+				_lastArabic = true;
+				_hasUbatAl = true;
+				cType = "R";
+				break;
+
+			case "WS":
+				cType = "N";
+				break;
+
+			case "CS":
+				if ( idx < 1 || ( idx + 1 ) >= types.length ||
+					( ( wType = resolvedTypes[idx - 1] ) !== "EN" && wType !== "AN" ) ||
+					( ( nType = types[idx + 1] ) !== "EN" && nType !== "AN" ) ) {
+					cType = "N";
+				} else if ( _lastArabic ) {
+					nType = "AN";
+				}
+				cType = ( nType === wType ) ? nType : "N";
+				break;
+
+			case "ES":
+				wType = idx > 0 ? resolvedTypes[idx - 1] : "B";
+				cType = wType === "EN" && ( idx + 1 ) < types.length && types[idx + 1] === "EN" ?
+						"EN" : "N";
+				break;
+
+			case "ET":
+				if ( idx > 0 && resolvedTypes[idx - 1] === "EN" ) {
+					cType = "EN";
+					break;
+				} else if ( _lastArabic ) {
+					cType = "N";
+					break;
+				}
+				i = idx + 1;
+				len = types.length;
+				while ( i < len && types[i] === "ET" ) {
+					i++;
+				}
+				if ( i < len && types[i] === "EN" ) {
+					cType = "EN";
+				} else {
+					cType = "N";
+				}
+				break;
+
+			case "NSM":
+				if ( _isInVisual && !_isInRtl ) { //V->L
+					len = types.length;
+					i = idx + 1;
+					while ( i < len && types[i] === "NSM" ) {
+						i++;
+					}
+					if ( i < len ) {
+						var c = chars[idx];
+						var rtlCandidate = ( c >= 0x0591 && c <= 0x08FF ) || c === 0xFB1E;
+						wType = types[i];
+						if ( rtlCandidate && ( wType === "R" || wType === "AL" ) ) {
+							cType = "R";
+							break;
+						}
+					}
+				}
+				if ( idx < 1 || ( wType = types[idx - 1] ) === "B" ) {
+					cType = "N";
+				} else {
+					cType = resolvedTypes[idx - 1];
+				}
+				break;
+
+			case "B":
+				_lastArabic = false;
+				_hasUbatB = true;
+				cType = _dir;
+				break;
+
+			case "S":
+				_hasUbatS = true;
+				cType = "N";
+				break;
+
+			case "LRE":
+			case "RLE":
+			case "LRO":
+			case "RLO":
+			case "PDF":
+				_lastArabic = false;
+				break;
+			case "BN":
+				cType = "N";
+				break;
+		}
+		return cType;
```

Comment:
Are the above rules available in CLDR somewhere? I'm wondering if they are perhaps available in XML only.